src/ParameterLists.f90

```diff
@@ -2413,40 +2422,55 @@ RECURSIVE FUNCTION validateReq_ParamType(thisParam,reqParams,prefix,isMatch) &
+              (validType == VALIDTYPE_VERIFYLIST)) THEN
             IF(ASSOCIATED(p%pdat)) &
-              isValid=validateReq_ParamType(thisParam,p%pdat,prefix,isMatch)
+              isValid=validateReq_ParamType(thisParam,p%pdat,prefix,validType,isMatch)
           ELSE
             IF(ASSOCIATED(p%pdat)) &
-              isValid=validateReq_ParamType(thisParam,p%pdat,prefix)
+              isValid=validateReq_ParamType(thisParam,p%pdat,prefix,validType,tmpbool)
           ENDIF
         TYPE IS(ParamType_List)
           !Loop over all parameters in the list and check each
           IF(ALLOCATED(p%pList)) THEN
             ntrue=0
             DO i=1,SIZE(p%pList)
-              IF(PRESENT(isMatch)) THEN
+              IF((validType == VALIDTYPE_VERIFYTEST) .OR. &
+                  (validType == VALIDTYPE_VERIFYLIST)) THEN
                 IF(validateReq_ParamType(thisParam,p%pList(i), &
-                  prefix//p%name//'->',isMatch)) ntrue=ntrue+1
+                  prefix//p%name//'->',validType,isMatch)) ntrue=ntrue+1
               ELSE
                 IF(validateReq_ParamType(thisParam,p%pList(i), &
-                  prefix//p%name//'->')) ntrue=ntrue+1
+                  prefix//p%name//'->',validType,tmpbool)) ntrue=ntrue+1
               ENDIF
             ENDDO
-            IF(ntrue == SIZE(p%pList)) isValid=.TRUE.
+            isValid=(ntrue == SIZE(p%pList))
           ELSE
             !The required list is not allocated, which means we do not
             !check any of it's possible subparameters, but we must at least
             !check that the list exists
             CALL thisParam%getParam(prefix//p%name,tmpParam)
             IF(.NOT.ASSOCIATED(tmpParam)) THEN
-              CALL eParams%raiseError(modName//'::'//myName// &
-                ' - Failed to locate required parameter "'//prefix// &
-                  p%name//'"!')
+              SELECTCASE(validType)
+                CASE(VALIDTYPE_VERIFYLIST,VALIDTYPE_VERIFYTEST)
+                  isMatch=.FALSE.
+                  CALL eParams%raiseError(modName//'::'//myName// &
+                      ' - When verifying that parameters are equal, the parameter "'// &
+                      prefix//p%name//'" was not found on both lists!')
+                CASE DEFAULT
+                  CALL eParams%raiseError(modName//'::'//myName// &
+                      ' - Failed to locate required parameter "'//prefix// &
+                      p%name//'"!')
+              ENDSELECT
             ELSE
               IF(SAME_TYPE_AS(tmpParam,p)) THEN
                 isValid=.TRUE.
-                IF(PRESENT(isMatch)) isMatch=match_ParamType(tmpParam,p,prefix)
+                SELECTCASE(validType)
+                  CASE(VALIDTYPE_VERIFYTEST)
+                    isMatch=isMatch .AND. matchTest_ParamType(tmpParam,p,prefix)
+                  CASE(VALIDTYPE_VERIFYLIST)
+                    isMatch=isMatch .AND. matchList_ParamType(tmpParam,p,prefix)
```

Comment:
This can be refactored as:
```
IF(validType == VALIDTYPE_VERIFYTEST .OR. validType == VALIDTYPE_VERIFYLIST) THEN
```
I don't think  you need a select construct here
?

Reply:
The selecttype is there because we need to call a different `match` subroutine depending on whether we are in a unit test (_VERIFYTEST) or in normal code execution (_VERIFYLIST).